feat: add clone complete endpoint and clone token to toc endpoint

richard015ar opened the pull request:
Issues: #3055 and #3057 .

These changes include changes in the API to add a new `clone_token` field in the `toc` API endpoint, which is the primary endpoint used during the clone routine.
That token is stored in the `wp_options` table and is used to receive notifications from the target book after the target network successfully cloned a book.
A new HTTP POST endpoint was added, `pressbooks/v2/clone/complete`, that expects `url` and `token` within the body to receive the successful clone notification and will store that in a new DB Table: `pressbooks_clones_complete`. 

Additionally, new interfaces were added following #3057.

arzola commented:
I added a couple of tests and I think is ready for final make up (css)

fdalcin commented on inc/cloner/class-cloner.php:
```diff
@@ -2067,4 +2076,52 @@ public static function isEnabled() {
+	 *
+	 * @return bool
+	 */
+	private function notifyCloneComplete(): bool {
```
This method feels more complex than it should? Do we need to split between local/remote books? Would be possible to use Guzzle to perform the request?

Just some thoughts

richard015ar replied:
For internal requests in WP we use https://developer.wordpress.org/reference/functions/rest_do_request/. 
Guzzle would be an option, but it is unnecessary in this case. Additionally, it will keep consistency in our other requests in the cloning routine, using the native WP_Request instead Guzzle.

Thanks for the idea and suggestion, Dalcin!

fdalcin commented on templates/admin/cloner/stats.blade.php:
```diff
@@ -0,0 +1,27 @@
+</div>
+@if($cloning_stats)
+<div class="pb-cloner stats wrap">
+	<h2 class="section-title">{{ __( 'Clones information', 'pressbooks' ) }}</h2>
```
I'd use `Clone information` instead, I don't know if it's correct English though

SteelWagstaff replied:
```suggestion
	<h2 class="section-title">{{ __( 'Clones made from this book:', 'pressbooks' ) }}</h2>

fdalcin reviewed (changes requested):
Code looks good, I left a couple of considerations. None of them are required though.

However, the `pb_clone_tokens` is not storing more than one token each time. I guess the problem is here because we're overriding the token list since we don't initialize the token list anywhere.

```
public function generateToken(): string {
	$token = md5( microtime() . wp_rand() );
	$this->tokens[ $token ] = time();
	$this->save();
	return $token;
}
```

richard015ar commented on inc/api/endpoints/controller/class-clonecomplete.php:
```diff
@@ -0,0 +1,89 @@
+	 */
+	public function __construct( CloneTokens $clone_tokens ) {
+		$this->clone_tokens = $clone_tokens;
+	}
```
Addressed. The `CloneComplete` was not injected, since it is not being used in some cases. It is used only in a particular case, so I think it is not a dependency, except for 1 method. I wrapped it up in a single line. Thanks for the suggestion!

richard015ar commented on inc/class-clonecomplete.php:
```diff
@@ -0,0 +1,72 @@
+
+class CloneComplete {
+
+	public static string $table = 'pressbooks_clones_complete';
```
I changed the name for one more clear, thanks!

richard015ar commented on inc/class-clonecomplete.php:
```diff
@@ -0,0 +1,72 @@
+		global $wpdb;
+		$table = static::$table;
+		$blog_id = get_current_blog_id();
+		return $wpdb->get_results( "SELECT * FROM {$wpdb->base_prefix}{$table} WHERE blog_id = $blog_id" ); // @codingStandardsIgnoreLine
```
I wrapped it. 
I instructed phpcs that ignores this line anyway, because of: 
``` 
Use placeholders and $wpdb->prepare(); found interpolated variable $table at "SELECT * FROM {$wpdb->base_prefix}{$table} WHERE blog_id = %d ORDER BY created_at DESC" - (WordPress.DB.PreparedSQL.InterpolatedNotPrepared)
```
Just like we did in different places within pressbooks.
Thanks!

richard015ar commented on inc/class-clonecomplete.php:
```diff
@@ -0,0 +1,72 @@
+				'blog_id' => $blog_id,
+				'target_book_name' => $target_book_name,
+				'target_book_url' => $target_book_url,
+				'created_at' => current_time( 'mysql' ),
```
addressed, thanks!

richard015ar commented on inc/api/endpoints/controller/class-toc.php:
```diff
@@ -266,9 +271,11 @@ public function get_item( $request ) {
-		$struct = $this->fixBookStructure( $struct, $has_permission );
+		$toc = $this->fixBookStructure( $struct, $has_permission );
+		$clone_tokens = new \Pressbooks\CloneTokens();
+		$toc['clone_token'] = $clone_tokens->generateToken();
```
Addressed, thanks!

richard015ar commented:
> 

Thanks Dalcin! I fixed the multiple tokens bug. I appreciate your revision and ideas! I addressed most of them.
I am assigning it to @arzola for revision.

arzola reviewed (approved):
LGTM I pass the torch to @SteelWagstaff to apply any kind of final CSS make up

SteelWagstaff reviewed (changes requested):
Pushed some changes to improve DOM & styles. Only thing I see now is that the hour:minutes:seconds are missing from the Date of clone value in the cloned books list:
![Screenshot from 2022-12-21 14-51-09](https://user-images.githubusercontent.com/13485451/209017861-4e66c937-ac3f-4382-8b0a-d7b747e73e89.png)
We should either:
1. get and display this information or
2. format the date so that it only shows Year-Month-Day

arzola commented:
@SteelWagstaff I fixed the date format I added Carbon date formatter so if you want we will be able to format the date in any other format you like for instance May 22, 2022, etc... but maybe is fine like it is because if not we would have to localize the date 😅 maybe a 2023 problem
